### PR TITLE
refactor XDPoS to accommodate multiple consensus engine versions

### PR DESCRIPTION
**Pull request by wjrjerome**
This PR shall not be merged until Adaptor change is ready and all CI tests are green

The next two steps are:
- Refact hooks
- Add unit tests

### CODE REVIEW

**Comment by liam-lai** [comment]
(c *XDPoS) -> (x *XDPoS), usually see people use the first letter

File: params/config.go [hunk 1]
```diff
@@ -169,13 +172,14 @@ func (c *CliqueConfig) String() string {
...
-	FoudationWalletAddr common.Address `json:"foudationWalletAddr"` // Foundation Address Wallet
-	SkipValidation      bool           //Skip Block Validation for testing purpose
+	Period                 uint64         `json:"period"`              // Number of seconds between blocks to enforce
+	Epoch                  uint64         `json:"epoch"`               // Epoch length to reset votes and checkpoint
+	Reward                 uint64         `json:"reward"`              // Block reward - unit Ether
+	RewardCheckpoint       uint64         `json:"rewardCheckpoint"`    // Checkpoint block for calculate rewards.
+	Gap                    uint64         `json:"gap"`                 // Gap time preparing for the next epoch
+	FoudationWalletAddr    common.Address `json:"foudationWalletAddr"` // Foundation Address Wallet
+	SkipValidation         bool           //Skip Block Validation for testing purpose
+	V2ConsensusBlockNumber *big.Int
```
Comment (liam-lai) [inline params/config.go:182]: maybe simply use int would work.

Comment (liam-lai) [reply params/config.go:182]: default value is 0 if not defined.

Comment (wjrjerome) [reply params/config.go:182]: the block number is big.int, that's why i try to keep it consistent.

File: params/config.go [hunk 2]
```diff
@@ -187,7 +191,10 @@ func (c *XDPoSConfig) String() string {
 ConsensusVersion will return the consensus version to use for the provided block number. The returned int represent its version
 TODO: It's a dummy value for now until the 2.0 consensus engine is fully implemented.
 */
-func BlockConsensusVersion(num *big.Int) int {
+func (c *XDPoSConfig) BlockConsensusVersion(num *big.Int) int {
+	if c.V2ConsensusBlockNumber != nil && num.Cmp(c.V2ConsensusBlockNumber) > 0 {
+		return 2
+	}
 	return 1
```
Comment (liam-lai) [inline params/config.go:198]: I think we can make these `1` and `2`, and const variable. 
Also, in the XDPoS.go, we can reassign variable. Then we don't need to always do package.variable
```

const (
	V1 = params.V1
)

```

File: params/config.go (hunk 1 above)
Comment (liam-lai) [reply params/config.go:182]: right, make sense.

**Comment by wjrjerome** [comment]
> (c *XDPoS) -> (x *XDPoS), usually see people use the first letter

The whole codebase is using c for type XDPoS.
I believe this stands for consensus which does make a bit sense to readability. However, I do agree with you we shall use the standard format for it. Does it make more sense to leave it

**Pull request closed without merging**